# Section 1 – Load Dataset

In [6]:
# -------------------------------------------------------
# SECTION 1: LOAD DATASET
# -------------------------------------------------------

import pandas as pd

train_df = pd.read_csv("/Users/shreekkanthkalyanasundaram/AIML/capstone_Q640/UNSW_NB15_training-set.csv")

print("="*60)
print("UNSW-NB15 Training Dataset Loaded")
print("="*60)

UNSW-NB15 Training Dataset Loaded


# Section 2 – Basic Exploratory Data Analysis

In [9]:
# -------------------------------------------------------
# SECTION 2: BASIC EXPLORATORY DATA ANALYSIS
# -------------------------------------------------------

print("\nDataset Shape")
print(train_df.shape)

print("\nColumn Names")
print(train_df.columns.tolist())

print("\nData Types")
print(train_df.dtypes)

print("\nFirst Five Records")
print(train_df.head())

print("\nSummary Statistics")
print(train_df.describe())


Dataset Shape
(175341, 45)

Column Names
['id', 'dur', 'proto', 'service', 'state', 'spkts', 'dpkts', 'sbytes', 'dbytes', 'rate', 'sttl', 'dttl', 'sload', 'dload', 'sloss', 'dloss', 'sinpkt', 'dinpkt', 'sjit', 'djit', 'swin', 'stcpb', 'dtcpb', 'dwin', 'tcprtt', 'synack', 'ackdat', 'smean', 'dmean', 'trans_depth', 'response_body_len', 'ct_srv_src', 'ct_state_ttl', 'ct_dst_ltm', 'ct_src_dport_ltm', 'ct_dst_sport_ltm', 'ct_dst_src_ltm', 'is_ftp_login', 'ct_ftp_cmd', 'ct_flw_http_mthd', 'ct_src_ltm', 'ct_srv_dst', 'is_sm_ips_ports', 'attack_cat', 'label']

Data Types
id                     int64
dur                  float64
proto                 object
service               object
state                 object
spkts                  int64
dpkts                  int64
sbytes                 int64
dbytes                 int64
rate                 float64
sttl                   int64
dttl                   int64
sload                float64
dload                float64
sloss                  

# Missing Values

In [10]:
print("\nMissing Values")

missing = train_df.isnull().sum()

print(missing[missing > 0])


Missing Values
Series([], dtype: int64)


# Class Distribution

In [11]:
print("\nBinary Target Distribution")

print(train_df["label"].value_counts())

print("\nPercentage Distribution")

print(train_df["label"].value_counts(normalize=True)*100)


Binary Target Distribution
label
1    119341
0     56000
Name: count, dtype: int64

Percentage Distribution
label
1    68.062233
0    31.937767
Name: proportion, dtype: float64


# Feature Count

In [12]:
predictors = train_df.drop(columns=["id","attack_cat","label"])

print("\nNumber of Predictor Variables")

print(len(predictors.columns))


Number of Predictor Variables
42


# Section 3 – Data Quality Assessment

In [14]:
# -------------------------------------------------------
# SECTION 3: DATA QUALITY ASSESSMENT
# -------------------------------------------------------

print("\nDuplicate Records")

print(train_df.duplicated().sum())

print("\nCategorical Features")

print(train_df.select_dtypes(include="object").columns.tolist())


Duplicate Records
0

Categorical Features
['proto', 'service', 'state', 'attack_cat']


# Section 4 – Sample Size Validation

# RQ1: Sample Size Justification

**Research Question 1**

> Which network telemetry features contribute most to accurately identifying network flow anomalies?

Before developing machine learning models, it is important to verify that the available dataset is sufficiently large to support reliable statistical analysis and model training.

Although the UNSW-NB15 dataset already provides predefined training and testing partitions, this study formally evaluates sample adequacy using two complementary approaches:

1. **Green's Rule** – a heuristic for estimating the minimum recommended sample size for multivariable analyses based on the number of predictor variables.
3. **Events Per Variable (EPV)** – a widely accepted criterion for classification models that measures the number of positive events available for each predictor variable.

Together, these measures provide statistical evidence that the dataset is sufficiently large for feature importance analysis and supervised machine learning.

In [14]:
import pandas as pd

# Load the training dataset
df = pd.read_csv("/Users/shreekkanthkalyanasundaram/AIML/capstone_Q640/UNSW_NB15_training-set.csv")

# Predictor variables (42 features)
X = df.drop(columns=["id", "attack_cat", "label"])

# Target variable (Binary classification)
y = df["label"]

# Sample statistics
n_samples = len(df)
n_predictors = X.shape[1]
n_events = (y == 1).sum()      # Anomalous network flows
n_non_events = (y == 0).sum()  # Normal network flows

# Green's Rule
green_rule = 50 + 8 * n_predictors

# Events Per Variable (EPV)
epv = n_events / n_predictors

print(f"Total Samples              : {n_samples}")
print(f"Predictor Variables        : {n_predictors}")
print(f"Positive Events (Anomaly)  : {n_events}")
print(f"Negative Events (Normal)   : {n_non_events}")
print(f"Green's Rule Minimum       : {green_rule}")
print(f"Events Per Variable (EPV)  : {epv:.2f}")

Total Samples              : 175341
Predictor Variables        : 42
Positive Events (Anomaly)  : 119341
Negative Events (Normal)   : 56000
Green's Rule Minimum       : 386
Events Per Variable (EPV)  : 2841.45


In [18]:
# ============================================================
# RESEARCH QUESTION 1
# Sample Size Justification
# ============================================================

"""
Research Question 1

Which network telemetry features contribute most to accurately
identifying operational anomalies?

Sample Size Methods:
1. Green's Rule
2. Events Per Variable (EPV)
"""

import pandas as pd

# ------------------------------------------------------------
# Load Training Dataset
# ------------------------------------------------------------

df = pd.read_csv("/Users/shreekkanthkalyanasundaram/AIML/capstone_Q640/UNSW_NB15_training-set.csv")

# ------------------------------------------------------------
# Predictor Variables and Target Variable
# ------------------------------------------------------------

X = df.drop(columns=["id", "attack_cat", "label"])
y = df["label"]

# ------------------------------------------------------------
# Dataset Statistics
# ------------------------------------------------------------

n_samples = len(df)
n_predictors = X.shape[1]

n_events = (y == 1).sum()
n_non_events = (y == 0).sum()

# ------------------------------------------------------------
# Green's Rule
# ------------------------------------------------------------

green_rule = 50 + 8 * n_predictors

# ------------------------------------------------------------
# Events Per Variable (EPV)
# ------------------------------------------------------------

epv = n_events / n_predictors

# ------------------------------------------------------------
# Results
# ------------------------------------------------------------

print("=" * 70)
print("RESEARCH QUESTION 1")
print("=" * 70)

print("\nResearch Question")
print("-" * 70)
print("Which network telemetry features contribute most to")
print("accurately identifying operational anomalies?")

print("\nDataset Statistics")
print("-" * 70)
print(f"Training Samples             : {n_samples:,}")
print(f"Predictor Variables          : {n_predictors}")
print(f"Positive Events (Anomalies)  : {n_events:,}")
print(f"Negative Events (Normal)     : {n_non_events:,}")

print("\nSample Size Assessment")
print("-" * 70)
print(f"Green's Rule Minimum         : {green_rule}")
print(f"Events Per Variable (EPV)    : {epv:.2f}")

print("\nConclusion")
print("-" * 70)
print("✓ Dataset exceeds Green's Rule requirement.")
print("✓ EPV is substantially greater than the recommended minimum.")
print("✓ Dataset is statistically adequate for feature importance")
print("  analysis and supervised machine learning.")

RESEARCH QUESTION 1

Research Question
----------------------------------------------------------------------
Which network telemetry features contribute most to
accurately identifying operational anomalies?

Dataset Statistics
----------------------------------------------------------------------
Training Samples             : 175,341
Predictor Variables          : 42
Positive Events (Anomalies)  : 119,341
Negative Events (Normal)     : 56,000

Sample Size Assessment
----------------------------------------------------------------------
Green's Rule Minimum         : 386
Events Per Variable (EPV)    : 2841.45

Conclusion
----------------------------------------------------------------------
✓ Dataset exceeds Green's Rule requirement.
✓ EPV is substantially greater than the recommended minimum.
✓ Dataset is statistically adequate for feature importance
  analysis and supervised machine learning.


## Green's Rule

Green (1991) proposed a heuristic for estimating the minimum sample size required for multivariable analyses.

The rule is expressed as:

n ≥ 50+8m

where

- **n** = minimum required sample size

- **m** = number of predictor variables

In this study:

- Predictor variables = **42**

Therefore,

n ≥ 50 + 8(42) = 386

The available training dataset contains **175,341 observations**, which is substantially larger than the minimum recommended sample size.

Although Green's Rule was originally proposed for multiple linear regression, it is included here as a supporting heuristic demonstrating that the dataset size greatly exceeds commonly accepted recommendations for multivariable analysis.

Green's Rule provides a supporting heuristic indicating that the available sample size substantially exceeds the minimum recommended for multivariable analysis. Because the primary analyses involve supervised machine learning classification, Green's Rule is complemented by the Events Per Variable (EPV) criterion, which serves as the principal sample adequacy measure for this study.

## Events Per Variable (EPV)

Because this study develops supervised machine learning classification models, Events Per Variable (EPV) provides a more appropriate measure of sample adequacy than Green's Rule.

Events Per Variable (EPV) is a widely accepted measure for assessing dataset adequacy in binary classification models.
                                   
                                   Number of Positive Events
     Events Per Variable (EPV) =  _________________________________
                                   Number of Predictor Variables
 

where

- Positive Events = observations labelled as anomalies (`label = 1`)
- Predictor Variables = independent variables used for model development

For this study:

- Positive Events = **119,341**
- Predictor Variables = **42**

Therefore,

               119341
       EPV =  ________ =2841.45
                42
       

Previous methodological research recommends a minimum EPV of **10**, while values of **20 or greater** are generally considered excellent for predictive modelling.

The calculated EPV of **2841.45** indicates that the dataset contains an exceptionally large number of anomaly observations relative to the number of predictors, providing strong evidence that the dataset is more than adequate for feature importance analysis and supervised machine learning.

Events are the number of observations belonging to the positive outcome class (label = 1), representing anomalous network flows. The EPV is computed by dividing the number of anomalous observations by the number of predictor variables used in model development.


# Research Question 2

## Can feature selection reduce telemetry processing overhead while maintaining sufficient anomaly detection performance for enterprise deployment?

### Why no separate sample size calculation?

Unlike Research Question 1, feature selection is a model development activity rather than a data collection problem. Therefore, no independent sample size formula is required.

The statistical adequacy of the dataset has already been established in RQ1 using:

- Green's Rule
- Events Per Variable (EPV)

Since feature selection uses the same training dataset, the previously established sample adequacy remains valid.

The feature selection process will be performed exclusively on the training dataset to prevent data leakage into the independent testing dataset.

In [19]:
# ============================================================
# RESEARCH QUESTION 2
# Dataset Adequacy for Feature Selection
# ============================================================

"""
Research Question 2

Can feature selection reduce telemetry processing overhead while
maintaining sufficient anomaly detection performance?

Sample Size Method:
Reuses the sample adequacy established in RQ1.
"""

# ------------------------------------------------------------
# Results
# ------------------------------------------------------------

print("=" * 70)
print("RESEARCH QUESTION 2")
print("=" * 70)

print("\nResearch Question")
print("-" * 70)
print("Can feature selection reduce telemetry processing")
print("overhead while maintaining anomaly detection performance?")

print("\nDataset Statistics")
print("-" * 70)
print(f"Training Samples             : {n_samples:,}")
print(f"Predictor Variables          : {n_predictors}")
print(f"Positive Events (Anomalies)  : {n_events:,}")
print(f"Negative Events (Normal)     : {n_non_events:,}")

print("\nSample Adequacy")
print("-" * 70)
print(f"Green's Rule Minimum         : {green_rule}")
print(f"Events Per Variable (EPV)    : {epv:.2f}")

print("\nConclusion")
print("-" * 70)
print("✓ Feature selection uses the same validated training dataset.")
print("✓ No additional sample size calculation is required.")
print("✓ Dataset adequacy established in RQ1 remains applicable.")
print("✓ Reduced feature sets will be validated using")
print("  cross-validation and the independent testing dataset.")

RESEARCH QUESTION 2

Research Question
----------------------------------------------------------------------
Can feature selection reduce telemetry processing
overhead while maintaining anomaly detection performance?

Dataset Statistics
----------------------------------------------------------------------
Training Samples             : 175,341
Predictor Variables          : 42
Positive Events (Anomalies)  : 119,341
Negative Events (Normal)     : 56,000

Sample Adequacy
----------------------------------------------------------------------
Green's Rule Minimum         : 386
Events Per Variable (EPV)    : 2841.45

Conclusion
----------------------------------------------------------------------
✓ Feature selection uses the same validated training dataset.
✓ No additional sample size calculation is required.
✓ Dataset adequacy established in RQ1 remains applicable.
✓ Reduced feature sets will be validated using
  cross-validation and the independent testing dataset.


## Interpretation

Feature selection does not require an independent sample size calculation because no additional observations are collected.

The same training dataset validated in Research Question 1 is used throughout the feature selection process.

Model robustness will instead be ensured through:

- Recursive Feature Elimination (RFE)
- Random Forest Feature Importance
- SHAP Feature Importance
- 10-fold Cross-Validation
- Independent evaluation using the UNSW-NB15 testing dataset

Therefore, the sample adequacy established in RQ1 is sufficient for Research Question 2.

RQ1 and RQ2 draw on the same underlying dataset and are related by nesting — since RQ2's reduced feature set (p=15) requires a smaller minimum sample than RQ1's full set (p=42), RQ1's adequacy determination extends to RQ2 by construction. RQ3 and RQ5, while using the same dataset, test fundamentally different statistical properties (comparative model performance and ROC/AUC precision, respectively) and therefore require independent adequacy checks. RQ4 requires entirely new primary data collection and is assessed independently.
